src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,399 @@
+then you can start migrating.
+Otherwise, use the **Resolvable** column to find
+null-safe releases, if they exist.
+
+For example, in the following output for a simple package,
+the green, checkmarked versions support null safety:
+
+![Output of dart pub outdated, showing PENDING](/null-safety/pub-outdated-output.png)
+
+Here are the steps to take before migrating this package to sound null safety:
+
+1. Wait for `examples_util` to be migrated.
+2. Update `pubspec.yaml` to use the latest compatible releases of
+   `pedantic`, `test`, and `examples_util`
+   (as listed in the **Resolvable** column).
+3. Run `dart pub upgrade`.
+4. Run `dart pub outdated --mode=null-safety` once more,
+   to make sure that all dependencies are null safe.
+
+{{ site.alert.info }}
+  **Why do all dependencies need to support null safety?**
+  When all direct dependencies support null safety,
+  you can _run your app_ with sound null safety.
+  When all dev dependencies support null safety,
+  you can _run tests_ with sound null safety.
+{{ site.alert.end }}
+
+
+## 2. Migrate {#step2-migrate}
+
+Most of the changes that your code needs to be null safe
+are easily predictable.
+For example, if a variable can be `null`,
+its type needs a `?` suffix.
+A named parameter that shouldn't be nullable
+needs to be marked `required`.
+
+The migration tool can make most of these changes,
+and you can add hints to customize the changes.
+Alternatively, you can
+[migrate your code by hand](#migrating-by-hand).
+
+{{ site.alert.tip }}
+  For additional help while migrating code, check the
+  [null safety FAQ][].
+{{ site.alert.end }}
+
+[null safety FAQ]: /null-safety/faq
+
+
+### Using the migration tool
+
+The migration tool takes a package of null-unsafe Dart code
+and converts it to null safety.
+
+Before running the tool,
+make sure you're using a 2.12 beta release (or later):
+
+```terminal
+$ dart --version
+Dart SDK version: 2.12.0-18.0.beta (beta)... [PENDING: check version #/string!]
+```
+
+Then update your `pubspec.yaml` file:
+
+* **Keep any null-safe package versions** you chose when you
+  [checked dependency status](#check-dependency-status) in step 1.
+* If your package's SDK constraints have a minimum version of 2.12 or higher:
+  1. **Set the minimum SDK version** back to a version before null safety.
+     ```yaml
+environment:
+    sdk: '>=2.9.0 <3.0.0'
+```
+  2. **Update dependencies:**
+     ```terminal
+$ dart pub get
+```
+
+Next, start the migration tool by running the `dart migrate` command
+in the directory that contains the package's `pubspec.yaml` file:
+
+```terminal
+$ dart migrate
+```
+
+If your package is ready to migrate,
+then the tool produces a line like the following:
+
+```terminal
+View the migration suggestions by visiting:
+
+  http://127.0.0.1:60278/Users/you/project/mypkg.console-simple?authToken=Xfz0jvpyeMI%3D
+```
+
+Visit that URL in a Chrome browser
+to see an interactive UI
+where you can guide the migration process:
```

Comment:
There is a bunch more details here for the suggested flow when using the tool; some of this might be relevant:
https://github.com/dart-lang/sdk/blob/master/pkg/nnbd_migration/README.md#using-the-tool

Reply:
I got some of that in there, but I didn't get `pub get` in there. Will do...

(I was surprised that comments got written in right away, btw. I thought they'd just be suggestions that wouldn't be applied until you clicked Apply. I ended up editing the source files directly a fair amount.)

Reply:
The `pub get` part shouldn't be needed here in step 2 as already covered that in step 1.

I think we can resolve this comment?

Reply:
I think the `pub get` is necessary because of moving the min SDK back to 2.9.

Reply:
That part shouldn't be needed either; we cherry picked this earlier today:
https://github.com/dart-lang/sdk/issues/44207

It still needs to be rolled to Flutter. After that we can both retest it.

Reply:
Updated, ptal

Reply:
Looks good!